src/sync/waitgroup.go

```diff
@@ -3,35 +3,62 @@ package sync
 import "internal/task"
 
 type WaitGroup struct {
-	counter uint
-	waiters task.Stack
+	futex task.Futex
 }
 
 func (wg *WaitGroup) Add(delta int) {
 	if delta > 0 {
 		// Check for overflow.
-		if uint(delta) > (^uint(0))-wg.counter {
-			panic("sync: WaitGroup counter overflowed")
-		}
+		for {
+			counter := wg.futex.Load()
+			if uint32(delta) > (^uint32(0))-counter {
+				panic("sync: WaitGroup counter overflowed")
+			}
 
-		// Add to the counter.
-		wg.counter += uint(delta)
-	} else {
-		// Check for underflow.
-		if uint(-delta) > wg.counter {
-			panic("sync: negative WaitGroup counter")
+			// Add to the counter.
+			if wg.futex.CompareAndSwap(counter, counter+uint32(delta)) {
+				// Successfully added.
+				return
+			}
 		}
+	} else {
```

Comment:
Remove `else`.  The `if` case lands in a `for {}` that either `panic`s or `return`s.

Reply:
I prefer to keep the structure as-is: while for things like fast paths and error return it makes sense to have an early return, in this case there are two equal cases: positive or negative/zero.
I can use a switch instead, though?

```go
switch {
case delta > 0:
    // positive
default:
    // negative (or zero)
}
```

Reply:
> I can use a switch instead, though?

:+1: IMO

Reply:
Ok, updated.